### ISSUE TITLE
Warn when Array.sort is used without a callback on non-string arrays

### ISSUE THREAD

**Issue by iansan5653** [microsoft/TypeScript#35551]
## Search Terms

array.sort, sort, numeric array sort, alphabetic array sort

## Suggestion

Give a warning when `Array.sort` is called without a callback on an array that is not of type `string[]`.

## Use Cases

By default `Array.sort()` sorts the array by the elements' UTF-16 code-point values after converting the array values to strings. This means that you'll get unexpected hidden results if you call `sort` on a numeric array (or any non-string array really). If you're intimately familiar with JavaScript, this is understood, but if you're coming from another language this would be surprising, unexpected, and hard to find.

## Examples

```ts
// Don't warn - this will give the expected result
["a", "d", "b", "f", "bat"].sort() // -> ["a", "b", "bat", "d", "f"]

// Warn - this is almost definitely a mistake
[0, 1, 2, 3, 20, 10].sort() // -> [0, 1, 10, 2, 20, 3]

// Warn - this is most likely not the expected result
[1, true, false, 0].sort() // -> [0, 1, false, true]

// Don't warn - a callback was provided to handle numbers
[0, 1, 2, 3, 20, 10].sort((a, b) => a - b) // -> [0, 1, 2, 3, 10, 20]
```

Another example is microsoft/vscode#86440, where this mistake found its way into production code.

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
 * _Not sure about this._
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by ilogico** [comment]
Something like this:
```typescript
interface Array<T> {
  sort(comparator: (a: T, b: T) => number): this;
  sort(this: Array<string>): this;
}
```
(the error message when the callback is omited is not very good though)
Technically, it's a breaking change. But I doubt it will break much legitimate code.
It will most likely reveal bugs.

**Comment by iansan5653** [comment]
@ilogico That's what I was thinking too, but I do think a more helpful error code would be needed because my first assumption on seeing that error would be that it's a bug with TypeScript and not with my code (since I know the comparator is optional).

**Comment by MartinJohns** [comment]
Related / duplicate of: #18286

**Comment by iansan5653** [comment]
Thanks @MartinJohns I'll close this and comment there
